arm.1
.TH arm 1 "27 August 2010"
.SH NAME
arm - Terminal Tor status monitor

.SH SYNOPSIS
arm [\fIOPTION\fR]

.SH DESCRIPTION
The anonymizing relay monitor (arm) is a terminal status monitor for Tor
relays, intended for command-line aficionados, ssh connections, and anyone
stuck with a tty terminal. This works much like top does for system usage,
providing real time statistics for:
  * bandwidth, cpu, and memory usage
  * relay's current configuration
  * logged events
  * connection details (ip, hostname, fingerprint, and consensus data)
  * etc

Defaults and interface properties are configurable via a user provided
configuration file (for an example see the provided \fBarmrc.sample\fR).
Releases and information are available at \fIhttp://www.atagar.com/arm\fR.

.SH OPTIONS
.TP
\fB\-i\fR, \fB\-\-interface [ADDRESS:]PORT\fR
tor control port arm should attach to (default is \fB127.0.0.1:9051\fR)

.TP
\fB\-c\fR, \fB\-\-config CONFIG_PATH\fR
user provided configuration file (default is \fB~/.arm/armrc\fR)

.TP
\fB\-d\fR, \fB\-\-debug\fR
writes all arm logs to ~/.arm/log

.TP
\fB\-b\fR, \fB\-\-blind\fR
disable connection lookups (netstat, lsof, and ss), dropping the parts of the
interface that rely on this information

.TP
\fB\-e\fR, \fB\-\-event EVENT_FLAGS\fR
flags for tor, arm, and torctl events to be logged (default is \fBN3\fR)

  d DEBUG      a ADDRMAP           k DESCCHANGED   s STREAM
  i INFO       f AUTHDIR_NEWDESCS  g GUARD         r STREAM_BW
  n NOTICE     h BUILDTIMEOUT_SET  l NEWCONSENSUS  t STATUS_CLIENT
  w WARN       b BW                m NEWDESC       u STATUS_GENERAL
  e ERR        c CIRC              p NS            v STATUS_SERVER
               j CLIENTS_SEEN      q ORCONN
    DINWE tor runlevel+            A All Events
    12345 arm runlevel+            X No Events
    67890 torctl runlevel+         U Unknown Events

.TP
\fB\-v\fR, \fB\-\-version\fR
provides version information

.TP
\fB\-h\fR, \fB\-\-help\fR
provides usage information

.SH FILES
.TP
\fB~/.armrc\fR
Your personal arm configuration file

.TP
\fB/usr/share/doc/arm/armrc.sample\fR
Sample armrc configuration file that documents all options

.SH AUTHOR
Written by Damian Johnson (atagar@torproject.org)



ChangeLog
CHANGE LOG

4/4/11 - version 1.4.2 (r24555)
This release chiefly consists of a fully reimplemented connection panel. Besides being a sane, maintainable implementation this includes numerous new features and improvements like full circuit paths, applications involved for local connections, and better type identification.

    * added: full rewrite of the connection panel, providing:
          o listing the full paths involved in active circuits
          o identification of socks, hidden service, and controller applications (arm, vidalia, polipo, etc)
          o identification of exit connections with the common usage for the port they're using
          o display of the local -> internal -> external address when room is available (original patch by Fabian Keil)
          o better accuracy and performance in identifying client and directory connections
          o marking the uptimes for initial connections (arm only tracks connection uptimes since starting, so these entries are just minimum durations)
          o lazily loading the initial IP -> fingerprint mappings to improve the startup time
          o using the circuit-status to disambiguating multiple relays on the same IP address
          o smarter space utilization, filling in smaller columns if there isn't room for higher priority but larger entries
          o connection details popup changes:
                + using the consensus exit policies rather than the longer descriptor versions when available
                + displaying connection details no longer freezes the rest of the display
                + detail panel uses the full screen width and is dynamically resizable
                + more resilient to missing descriptors
    * change: hiding most tor config values by default (idea by arma)
    * change: dropping warning suggesting that users set the FetchUselessDescriptors option (suggestion by Sebastian and others)
    * change: always starting the bandwidth field from zero rather than using the state file total, which only contains the last day's worth of data (thanks to guilhem)
    * change: suggesting authentication and giving steps for it in the readme (suggestion by Sebastian)
    * change: caching config display lines, which reduces the CPU usage when scrolling by around 40%
    * change: added summaries for the remaining tor configuration options
    * change: using a dedicated enum class rather than tuple sets
    * fix: torrc validation requires 'GETINFO config-text' which was introduced in Tor verison 0.2.2.7 (caught by Sjon, talion, and torland, https://trac.torproject.org/projects/tor/ticket/2501)
    * fix: off-by-one issue with the displayed line numbers for torrc errors (caught by Sjon)
    * fix: bin function wasn't available before python 2.6 (caught by Paul Menzel)
    * fix: mis-parsing family entries when there's no entry after the comma (caught by StrangeCharm, https://trac.torproject.org/projects/tor/ticket/2414)
    * fix: preventing SOCKS and CONTROL connections from being expanded (patch by Fabian Keil)
    * fix: disabling name resolution for application queries to avoid leaking to resolvers (patch by Fabian Keil)
    * fix: reversing src and dst addresses of SOCKS and CONTROL connections (caught by Fabian Keil)
    * fix: changing the 'APPLICATION' type to 'SOCKS' since the previous label was too long (caught by Fabian Keil)
    * fix: crashing issue from unknown relay nicknames (caught by krkhan)
    * fix: concurrency bug occasionally causing "syshook" stacktraces when shutting down
    * fix: header panel displayed the wrong IP address if it changed since we first started (https://trac.torproject.org/projects/tor/ticket/2776)
    * fix: unchecked OSError could cause us to crash when making directories (for instance if there was a permissions issue)
    * fix: the availability check for bsd resolvers was broken, probably causing resolution to fail for a few seconds on that platform
    * fix: dropping the pointless 'Log notice stdout' entry provided by config-text queries (https://trac.torproject.org/projects/tor/ticket/2362)
    * fix: taking DirServer and AlternateDirAuthority into account when determining the directory authorities
    * fix: consuming a little extra space in the connection panel when scrollbars aren't visible
    * fix: dropping the deprecated 'features.config.descriptions.persistPath' config option
    * fix: failed connection attempts to the control port were generating zombie connections (https://trac.torproject.org/projects/tor/ticket/2812)
    * fix: concurrency bug in joining on the TorCtl thread when tor shut down
    * fix: the 'startup.dataDirectory' config option was being ignored
    * fix: recognizing the proper private ip ranges of the 172.* block
    * fix: missing 'is default' option from config sort ordering
    * fix (4/4/11, r24562): hidden service parsing issue when there's multiple spaces in the HiddenServicePort opition (caught by Nicolas Pouillard)
    * fix (4/6/11, r24570): missing new connection components from installations (caught by Anthony Basile)
    * fix (4/13/11, r24613): failed requests for our flags cause a syntax error (caught by qbi)

1/7/11 - version 1.4.1 (r24054)
Platform specific enhancements including BSD compatibility and vastly improved performance on Linux.

    * added: querying the proc contents when able for tor resource and connection samplings to *greatly* reduce arm's resource usage (many thanks to psutil)
    * added: vastly improved BSD compatibility, thanks to patches by Fabian Keil
          o pid resolution via pgrep (all platforms) and sockstat (bsd only)
          o connection resolution via sockstat (all platforms) and procstat (bsd only)
          o autodetecting the path prefix for FreeBSD jails
    * added: displaying summaries of the options on the configuration panel (idea by Sebastian)
    * added: arm cpu usage to the header panel and logs (with an estimate for system call usages)

arm.py
#!/usr/bin/env python
# arm.py -- Terminal status monitor for Tor relays.
# Released under the GPL v3 (http://www.gnu.org/licenses/gpl.html)

"""
Command line application for monitoring Tor relays, providing real time status
information. This is the starter for the application, handling and validating
command line parameters.
"""

import sys
import socket
import getopt
import getpass

from TorCtl import TorCtl
from TorCtl import TorUtil

from interface import controller
from interface import logPanel

VERSION = "1.3.1"
LAST_MODIFIED = "Feb 7, 2010"

DEFAULT_CONTROL_ADDR = "127.0.0.1"
DEFAULT_CONTROL_PORT = 9051
DEFAULT_LOGGED_EVENTS = "N3" # tor and arm NOTICE, WARN, and ERR events

OPT = "i:p:be:vh"
OPT_EXPANDED = ["interface=", "password=", "blind", "event=", "version", "help"]
HELP_TEXT = """Usage arm [OPTION]
Terminal status monitor for Tor relays.

  -i, --interface [ADDRESS:]PORT  change control interface from %s:%i
  -p, --password PASSWORD         authenticate using password (skip prompt)
  -b, --blind                     disable connection lookups
  -e, --event EVENT_FLAGS         event types in message log  (default: %s)
%s
  -v, --version                   provides version information
  -h, --help                      presents this help

Example:
arm -b -i 1643          hide connection data, attaching to control port 1643
arm -e=we -p=nemesis    use password 'nemesis' with 'WARN'/'ERR' events
""" % (DEFAULT_CONTROL_ADDR, DEFAULT_CONTROL_PORT, DEFAULT_LOGGED_EVENTS, logPanel.EVENT_LISTING)

def isValidIpAddr(ipStr):
  """
  Returns true if input is a valid IPv4 address, false otherwise.
  """
  
  for i in range(4):
    if i < 3:
      divIndex = ipStr.find(".")
      if divIndex == -1: return False # expected a period to be valid
      octetStr = ipStr[:divIndex]
      ipStr = ipStr[divIndex + 1:]
    else:
      octetStr = ipStr
    
    try:
      octet = int(octetStr)
      if not octet >= 0 or not octet <= 255: return False
    except ValueError:
      # address value isn't an integer
      return False
  
  return True

if __name__ == '__main__':
  controlAddr = DEFAULT_CONTROL_ADDR     # controller interface IP address
  controlPort = DEFAULT_CONTROL_PORT     # controller interface port
  authPassword = ""                      # authentication password (prompts if unset and needed)
  isBlindMode = False                    # allows connection lookups to be disabled
  loggedEvents = DEFAULT_LOGGED_EVENTS   # flags for event types in message log
  
  # parses user input, noting any issues
  try:
    opts, args = getopt.getopt(sys.argv[1:], OPT, OPT_EXPANDED)
  except getopt.GetoptError, exc:
    print str(exc) + " (for usage provide --help)"
    sys.exit()
  
  for opt, arg in opts:
    if opt in ("-i", "--interface"):
      # defines control interface address/port
      try:
        divIndex = arg.find(":")
        
        if divIndex == -1:
          controlPort = int(arg)
        else:
          controlAddr = arg[0:divIndex]
          controlPort = int(arg[divIndex + 1:])
        
        # validates that input is a valid ip address and port
        if divIndex != -1 and not isValidIpAddr(controlAddr):
          raise AssertionError("'%s' isn't a valid IP address" % controlAddr)
        elif controlPort < 0 or controlPort > 65535:
          raise AssertionError("'%s' isn't a valid port number (ports range 0-65535)" % controlPort)
      except ValueError:
        print "'%s' isn't a valid port number" % arg
        sys.exit()
      except AssertionError, exc:
        print exc
        sys.exit()
    elif opt in ("-p", "--password"): authPassword = arg    # sets authentication password
    elif opt in ("-b", "--blind"): isBlindMode = True       # prevents connection lookups
    elif opt in ("-e", "--event"): loggedEvents = arg       # set event flags
    elif opt in ("-v", "--version"):
      print "arm version %s (released %s)\n" % (VERSION, LAST_MODIFIED)
      sys.exit()
    elif opt in ("-h", "--help"):
      print HELP_TEXT
      sys.exit()
  
  # validates and expands log event flags
  try:
    expandedEvents = logPanel.expandEvents(loggedEvents)
  except ValueError, exc:
    for flag in str(exc):
      print "Unrecognized event flag: %s" % flag
    sys.exit()
  
  # temporarily disables TorCtl logging to prevent issues from going to stdout while starting
  TorUtil.loglevel = "NONE"
  
  # attempts to open a socket to the tor server
  try:
    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    s.connect((controlAddr, controlPort))
    conn = TorCtl.Connection(s)
  except socket.error, exc:
    if str(exc) == "[Errno 111] Connection refused":
      # most common case - tor control port isn't available
      print "Connection refused. Is the ControlPort enabled?"
    else:
      # less common issue - provide exc message
      print "Failed to establish socket: %s" % exc
    
    sys.exit()
  
  # check PROTOCOLINFO for authentication type
  try:
    authInfo = conn.sendAndRecv("PROTOCOLINFO\r\n")[1][1]
  except TorCtl.ErrorReply, exc:
    print "Unable to query PROTOCOLINFO for authentication type: %s" % exc
    sys.exit()
  
  try:
    if authInfo.startswith("AUTH METHODS=NULL"):
      # no authentication required
      conn.authenticate("")
    elif authInfo.startswith("AUTH METHODS=HASHEDPASSWORD"):
      # password authentication, promts for password if it wasn't provided
      if not authPassword: authPassword = getpass.getpass()
      conn.authenticate(authPassword)
    elif authInfo.startswith("AUTH METHODS=COOKIE"):
      # cookie authtication, parses path to authentication cookie
      start = authInfo.find("COOKIEFILE=\"") + 12
      end = authInfo[start:].find("\"")
      authCookiePath = authInfo[start:start + end]
      
      try:
        authCookie = open(authCookiePath, "r")
        conn.authenticate_cookie(authCookie)
        authCookie.close()
      except IOError, exc:
        # cleaner message for common errors
        issue = None
        if str(exc).startswith("[Errno 13] Permission denied"): issue = "permission denied"
        elif str(exc).startswith("[Errno 2] No such file or directory"): issue = "file doesn't exist"
        
        # if problem's recognized give concise message, otherwise print exception string
        if issue: print "Failed to read authentication cookie (%s): %s" % (issue, authCookiePath)
        else: print "Failed to read authentication cookie: %s" % exc
        
        sys.exit()
    else:
      # authentication type unrecognized (probably a new addition to the controlSpec)
      print "Unrecognized authentication type: %s" % authInfo
      sys.exit()
  except TorCtl.ErrorReply, exc:
    # authentication failed
    issue = str(exc)
    if str(exc).startswith("515 Authentication failed: Password did not match"): issue = "password incorrect"
    if str(exc) == "515 Authentication failed: Wrong length on authentication cookie.": issue = "cookie value incorrect"
    
    print "Unable to authenticate: %s" % issue
    sys.exit()
  
  controller.startTorMonitor(conn, expandedEvents, isBlindMode)
  conn.close()

